transport/http2_client.go

```diff
@@ -922,26 +921,48 @@ func (t *http2Client) handleSettings(f *http2.SettingsFrame, isFirst bool) {
 	if f.IsAck() {
 		return
 	}
-	var ss []http2.Setting
+	var rs []http2.Setting
+	var ps []http2.Setting
 	isMaxConcurrentStreamsMissing := true
 	f.ForeachSetting(func(s http2.Setting) error {
 		if s.ID == http2.SettingMaxConcurrentStreams {
 			isMaxConcurrentStreamsMissing = false
 		}
-		ss = append(ss, s)
+		if t.isRestrictive(s) {
+			rs = append(rs, s)
+		} else {
+			ps = append(ps, s)
+		}
 		return nil
 	})
 	if isFirst && isMaxConcurrentStreamsMissing {
```

Comment:
Optional, maybe for a later refactor:

The code might be cleaner with two implementations: `handleSettings` and `handleInitialSettings`.  `handleInitialSettings` doesn't need to check restrictive/permissive -- just apply all the settings and write the ack.  Then the special-case code checking for max concurrent streams is removed from the ongoing `handleSettings` function.

Reply:
The same logic, acking after applying restrictive settings and before applying permissive settings, holds for handleInitialSettings. The client doesn't wait for the server to send initial settings before starting streams. 
We'll still need to check if `http2.SettingMaxConcurrentStreams` is present or not in the initial settings. So we will only be able to save one conditional check by creating this other function.